bentoml/_internal/server/service_app.py

```diff
@@ -313,17 +315,37 @@ async def api_func(
             # handle_request may raise 4xx or 5xx exception.
             try:
                 input_data = await api.input.from_http_request(request)
+                response = await api.output.init_http_response()
+                ctx = None
                 if asyncio.iscoroutinefunction(api.func):
                     if isinstance(api.input, Multipart):
+                        if api.needs_ctx:
+                            ctx = Context.from_http(request, response)
+                            input_data[api.ctx_param] = ctx
                         output = await api.func(**input_data)
                     else:
-                        output = await api.func(input_data)
+                        if api.needs_ctx:
+                            ctx = Context.from_http(request, response)
+                            output = await api.func(input_data, ctx)
+                        else:
+                            output = await api.func(input_data)
                 else:
                     if isinstance(api.input, Multipart):
+                        if api.needs_ctx:
+                            ctx = Context.from_http(request, response)
+                            input_data[api.ctx_param] = ctx
                         output: t.Any = await run_in_threadpool(api.func, **input_data)
                     else:
-                        output: t.Any = await run_in_threadpool(api.func, input_data)
-                response = await api.output.to_http_response(output)
+                        if api.needs_ctx:
+                            ctx = Context.from_http(request, response)
+                            output = await run_in_threadpool(
+                                api.func(input_data, ctx=ctx)
+                            )
+                        else:
+                            output = await run_in_threadpool(api.func, input_data)
+
+                await api.output.finalize_http_response(response, output)
+                finalize_http_response(response)
```

Comment:
it looks like we are calling this function both here and in the `api.output.finalize_http_response` above, do we need both?

Reply:
I think this is unnecessary since I'm calling that when I need to. Honestly it should be renamed to something like `set_content_length` now too.